src/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs

```diff
@@ -205,25 +205,31 @@ private void SetCollectionObjectWithCollectionMember(ref object collection, Coll
                     collection = ReflectionCreateObject(collectionType);
                 }
 
-                if (typeof(IList).IsAssignableFrom(collectionType))
+                AddObjectsIntoTargetCollection(collection, collectionMember, collectionType);
+            }
+        }
+
+        private static void AddObjectsIntoTargetCollection(object targetCollection, List<object> sourceCollection, Type targetCollectionType)
+        {
+            var targetList = targetCollection as IList;
+            if (targetList != null)
+            {
+                foreach (var item in sourceCollection)
                 {
-                    foreach (var item in collectionMember)
-                    {
-                        ((IList)collection).Add(item);
-                    }
+                    targetList.Add(item);
                 }
-                else
+            }
+            else
+            {
+                foreach (var item in sourceCollection)
                 {
-                    foreach (var item in collectionMember)
+                    MethodInfo addMethod = targetCollectionType.GetMethod("Add");
+                    if (addMethod == null)
                     {
-                        MethodInfo addMethod = collectionType.GetMethod("Add");
-                        if (addMethod == null)
-                        {
-                            throw new InvalidOperationException("(addMethod == null)");
-                        }
-
-                        addMethod.Invoke(collection, new object[] { item });
+                        throw new InvalidOperationException("addMethod == null");
                     }
+
+                    addMethod.Invoke(targetCollection, new object[] { item });
```

Comment:
Can the `object[]` be allocated once, at the beginning of the loop, and then reused rather than re-allocated for each item?